src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerDiagnosticListener.cs

```diff
@@ -115,16 +111,54 @@ public void OnStartActivity(Activity activity, object payload)
                 return;
             }
 
+            bool isNullActivitySource = string.IsNullOrEmpty(activity.Source.Name);
+
             // Propagate context irrespective of sampling decision
             var textMapPropagator = Propagators.DefaultTextMapPropagator;
             if (textMapPropagator is not TraceContextPropagator)
             {
-                textMapPropagator.Inject(new PropagationContext(activity.Context, Baggage.Current), request, HttpRequestMessageContextPropagation.HeaderValueSetter);
+                var context = activity.Context;
+
+                // Note: isNullActivitySource is true for all .NET 6 and prior
+                // spans. isNullActivitySource is true for .NET 7+ only when
+                // sample decision was drop.
+                Debug.Assert(
+                    isNullActivitySource || IsNet7OrGreater,
+                    "activity source was invalid");
+
+                // Note: When sample decision is drop...
+                //  * On .NET 7 activity has empty source and is NOT updated by
+                //    the legacy logic. Recorded is just mirrored from the
+                //    parent.
+                //
+                //  * On .NET 6 activity is updated by the legacy logic and Recorded can be trusted.
+                if (isNullActivitySource
```

Comment:
How do we know if the parent was Recorded?

Reply:
Can you give me more information about why are you asking? Because it doesn't really matter if the parent was recorded or not. What we know in this code is whether or not the span created by DiagnosticsHandler is a stub. If it is a stub, and we have a parent, that means the stub is not root. In that case we send the root (recorded or not).

Reply:
Because of the comment here regarding broken traces: https://github.com/open-telemetry/opentelemetry-dotnet/blob/2fb03a8dcfc3c07038c2ae45effb6977fdacbf42/src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerDiagnosticListener.cs#L139-L146.

Reply:
So ya it doesn't check if the parent is recorded. I don't think it needs to (I could be wrong). Just knowing we have a parent and we are a stub is enough info. I had at one point this walking the parent tree if there are multiple, but removed it in the end because I don't think that is needed either. Basically this propagates the trace as if the stub didn't exist 😄

Reply:
I am missing how that helps with broken trace :). Assuming it means broken end-to-end view in backend because of missing span.

Consider the following sequence in a trace

1. Incoming http request -- `Span 1` -- `Recorded = true`
2. Internal operation -- `Span 2` -  `Recorded = false`
3. Http Call -- `Span 3` - `Recorded = false`

Considering this change for propagation, we use parent of `Span 3` for propagation i.e `Span 2`. This still results in broken trace as `Span 2` is also not recorded.

Reply:
I don't think it is common for the root to be sampled and the children dropped. This is what I'm thinking about:

1. Incoming http request -- Span 1 -- Recorded = false. The root span has not been sampled, but Activity is still created. Explicitly for propagation! In this scenario the correct behavior is all children are dropped and the root is propagated.

2. Some logs are written correlated to the root span.

3. Internal operation -- Span 2 - Recorded = false. No Activity is created because sampler returns drop.

4. Wcf call -- Span 3 - Recorded = false. No Activity is created because sampler returns drop. Span 1 is propagated. In remote service new root is created with a parent of Span 1. Some logs are written correlated to this new remote root.

5. Some logs are written correlated to the root span.

6. Http Call -- Span 4 - Recorded = false. Activity is created as a stub by the runtime. Span 4 is propagated. In remote service new root is created with a parent of Span 4. Some logs are written correlated to this new remote root.

In the remote service activated by the Http Call (6) the incoming trace is "broken" in that it has a span that shouldn't have been created and won't be found anywhere. Will that cause problems? Not sure. But why not have it be correct if we can help it? 😄

Reply:
> In the remote service activated by the Http Call (6) the incoming trace is "broken" in that it has a span that shouldn't have been created and won't be found anywhere.

What do we know about why this span is being created? Are you suggesting this is a bug in the runtime when you say it shouldn't have been created?

I haven't looked at the runtime's code yet, but from the comments in our code it sounds like this span is created by the mere presence of a DiagnosticListener. Let's say there were another DiagnosticListener besides ours: in this scenario should we propagate the context of this missing span?

Reply:
> What do we know about why this span is being created? Are you suggesting this is a bug in the runtime when you say it shouldn't have been created?

[Prior to .NET 7 an Activity was always created directly (ActivitySource was not used)](https://github.com/dotnet/runtime/blob/5a400c212afdf8e675c9a1d38442e6d2f19f7b74/src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs#L101). Going forward, [an ActivitySource is used](https://github.com/dotnet/runtime/blob/550605cc93d9b903c4fbaf8f1f83e387d2f79dbe/src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs#L61) however is we drop the Activity, [it is still created as an extra step](https://github.com/dotnet/runtime/blob/550605cc93d9b903c4fbaf8f1f83e387d2f79dbe/src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs#L68). The reason is to preserve the old behavior where there was no sampling and an Activity always existed.

> Let's say there were another DiagnosticListener besides ours: in this scenario should we propagate the context of this missing span?

Not sure. I don't think so? If users have the OTel SDK configured they are probably interested (maybe hyper-interested) in accurate propagation. It might be fine to send the junk, but what's the harm in making it correct? Tiny perf hit to do it 🤷

Reply:
Related #3290.

> Going forward, [an ActivitySource is used](https://github.com/dotnet/runtime/blob/550605cc93d9b903c4fbaf8f1f83e387d2f79dbe/src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs#L61) however is we drop the Activity, [it is still created as an extra step](https://github.com/dotnet/runtime/blob/550605cc93d9b903c4fbaf8f1f83e387d2f79dbe/src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs#L68).

Trying to put aside some of the additional complexities of .NET's OTel implementation for a moment... From the standpoint of the spec, if some library were to:

```csharp
var tracer = GlobalTracerProvider.GetTracer("MyTracer");
var span = tracer.StartSpan("MySpan");
GlobalPropagator.Inject(span.Context, carrier, setter);
```

Per the spec, we would expect the span's context to be propagated regardless of whether it was sampled or dropped. Not the span's parent's context.

In some sense, isn't this "legacy" behavior of HttpClient kind of fixing #3290 for this isolated use case.

Reply:
This is an interesting case you brought up. If there is a root non-sampled span running, then this executes like...

```csharp
var span = tracer.StartSpan("MySpan"); // Returns NoopInstance when sample is drop
GlobalPropagator.Inject(span.Context, carrier, setter); // Injects default context from NoopInstance
```

We would end up injecting junk and breaking propagation in this case? Different issue 😄

> In some sense, isn't this "legacy" behavior of HttpClient kind of fixing https://github.com/open-telemetry/opentelemetry-dotnet/issues/3290 for this isolated use case.

Accidentally, yes. But I don't think the fix is to ask everyone to create the activity when it comes back `null`. The fix would probably to have `ActivitySource` do that?

I feel like the SDK's job is to be consistent and not have weird one-off/exception cases but I don't feel passionately about this particular thing 🤣 I'm going to revert this part and just fix the .NET 7 propagator break.

Reply:
> The fix would probably to have ActivitySource do that?

Yes that would be the real solution. 100% agree the httpclient behavior is an accidental fix that should not become the expected pattern.

Reply:
After I sent that I had the thought... if we returned `ActivitySamplingResult.PropagationData` where we return `ActivitySamplingResult.None` today we would essentially tell `ActivitySource` to always create the spans. Might take care of #3290.